src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -152,7 +165,8 @@ protected TransportRequestOptions transportOptions() {
     }
 
     protected boolean retryPrimaryException(Throwable e) {
-        return TransportActions.isShardNotAvailableException(e);
+        return e.getClass() == RetryOnPrimaryException.class
```

Comment:
Actually IllegalIndexShardStateException seems to be about shards not being in the right shard state (STARTED, RECOVERING, ...) which does not seem to apply here